tests/legacy-cli/e2e/tests/build/server-rendering/server-routes-preload-links.ts

```diff
@@ -0,0 +1,127 @@
+import assert from 'node:assert';
+import { writeFile } from '../../../utils/fs';
+import { execAndWaitForOutputToMatch, ng, noSilentNg, silentNg } from '../../../utils/process';
+import { installWorkspacePackages, uninstallPackage } from '../../../utils/packages';
+import { ngServe, updateJsonFile, useSha } from '../../../utils/project';
+import { getGlobalVariable } from '../../../utils/env';
+import { findFreePort } from '../../../utils/network';
+
+export default async function () {
+  assert(
+    getGlobalVariable('argv')['esbuild'],
+    'This test should not be called in the Webpack suite.',
+  );
+
+  // Forcibly remove in case another test doesn't clean itself up.
+  await uninstallPackage('@angular/ssr');
+  await ng('add', '@angular/ssr', '--server-routing', '--skip-confirmation', '--skip-install');
+  await useSha();
+  await installWorkspacePackages();
+
+  // Add routes
+  await writeFile(
+    'src/app/app.routes.ts',
+    `
+  import { Routes } from '@angular/router';
+
+  export const routes: Routes = [
+    {
+      path: '',
+      loadComponent: () => import('./home/home.component').then(c => c.HomeComponent),
+    },
+    {
+      path: 'ssg',
+      loadComponent: () => import('./ssg/ssg.component').then(c => c.SsgComponent),
+    },
+    {
+      path: 'ssr',
+      loadComponent: () => import('./ssr/ssr.component').then(c => c.SsrComponent),
+    },
+    {
+      path: 'csr',
+      loadComponent: () => import('./csr/csr.component').then(c => c.CsrComponent),
+    },
+  ];
+  `,
+  );
+
+  // Add server routing
+  await writeFile(
+    'src/app/app.routes.server.ts',
+    `
+  import { RenderMode, ServerRoute } from '@angular/ssr';
+
+  export const serverRoutes: ServerRoute[] = [
+    {
+      path: 'ssr',
+      renderMode: RenderMode.Server,
+    },
+    {
+      path: 'csr',
+      renderMode: RenderMode.Client,
+    },
+    {
+      path: '**',
+      renderMode: RenderMode.Prerender,
+    },
+  ];
+  `,
+  );
+
+  await updateJsonFile('angular.json', (workspaceJson) => {
+    const appProject = workspaceJson.projects['test-project'];
+    appProject.architect['build'].options.namedChunks = true;
+  });
+
+  // Generate components for the above routes
+  const componentNames: string[] = ['home', 'ssg', 'csr', 'ssr'];
+
+  for (const componentName of componentNames) {
+    await silentNg('generate', 'component', componentName);
+  }
+
+  // Test both vite and `ng build`
+  await runTests(await ngServe());
+
+  await noSilentNg('build', '--output-mode=server');
+  await runTests(await spawnServer());
+}
+
+const RESPONSE_EXPECTS: Record<string, RegExp> = {
+  '/': /<link rel="modulepreload" href="(home\.component-[a-zA-Z0-9]{8}\.js)">/,
+  '/ssg': /<link rel="modulepreload" href="(ssg\.component-[a-zA-Z0-9]{8}\.js)">/,
+  '/ssr': /<link rel="modulepreload" href="(ssr\.component-[a-zA-Z0-9]{8}\.js)">/,
+  '/csr': /<link rel="modulepreload" href="(csr\.component-[a-zA-Z0-9]{8}\.js)">/,
+};
```

Comment:
Consider: Would it be worth asserting that other the other preloads were _not_ included? We wouldn't want `/` to preload ssg.component-*.js`.